src/main/java/com/nutomic/syncthingandroid/preferences/WifiSsidPreference.java

```diff
@@ -0,0 +1,249 @@
+import java.util.List;
+import java.util.Set;
+
+/**
+ * Dialog preference which allows the user to select on which WiFi networks (based on SSID)
+ * syncing should be allowed.
+ *
+ * Setting can be "All networks" (none selected), or selecting individual networks.
+ *
+ * Due to restrictions in Android, it is possible/likely, that the list of saved WiFi networks
+ * cannot be retrieved if the WiFi is turned off. In this case, an explanation is shown.
+ *
+ * The preference is stored as Set&lt;String&gt; where <code>null</code> represents
+ * "all networks allowed", and a set represents a restriction to the selected SSIDs.
+ *
+ * SSIDs are formatted according to the naming convention of WifiManager, i.e. they have the
+ * surrounding double-quotes (") for UTF-8 names, or they are hex strings (if not quoted).
+ */
+public class WifiSsidPreference extends DialogPreference {
+
+    /**
+     * List containing either a notification that WiFi is disabled, or the list of saved
+     * networks.
+     */
+    private ListView mListWifi;
+    /**
+     * The configured WiFi networks, or null, if WiFi is not available.
+     */
+    private WifiConfiguration[] mConfiguredNetworks;
+    /**
+     * The SSID of the networks that have been enabled (in selective mode)
+     */
+    private Set<String> mSelectedSsids;
+    /**
+     * Adapter between the data above and the ListView
+     */
+    private WifiListAdapter listAdapter;
+
+    public WifiSsidPreference(Context context, AttributeSet attrs) {
+        super(context, attrs);
+        setDialogLayoutResource(R.layout.preference_wifi_ssids);
+    }
+
+    @Override
+    protected View onCreateDialogView() {
+        View view = super.onCreateDialogView();
+        mListWifi = (ListView) view.findViewById(R.id.listWifi);
+        return view;
+    }
+
+    /**
+     * This is where the data is actually bound.
+     */
+    @Override
+    protected void onBindDialogView(View view) {
+        super.onBindDialogView(view);
+
+        if (isPersistent()) {
+            // currently persisted values are pre-selected in the list
+            String key = getKey();
+            Set<String> ssids = getSharedPreferences().getStringSet(key, null);
+            mSelectedSsids = ssids != null ? new HashSet<>(ssids) : new HashSet<String>();
+        } else {
+            mSelectedSsids = new HashSet<>(); // if non-persistent, nothing is pre-selected
+        }
+
+        mConfiguredNetworks = loadConfiguredNetworksSorted();
+        listAdapter = new WifiListAdapter();
+        mListWifi.setAdapter(listAdapter);
+    }
+
+    /**
+     * Load the configured WiFi networks, sort them by SSID.
+     *
+     * @return a sorted array of WifiConfiguration, or null, if data cannot be retrieved
+     */
+    private WifiConfiguration[] loadConfiguredNetworksSorted() {
+        WifiManager wifiManager = (WifiManager) getContext().getSystemService(Context.WIFI_SERVICE);
+        if (wifiManager != null) {
```

Comment:
According to JavaDoc it may be `null` if:
> The service or null if the class is not a supported system service. 

There are devices without WiFi but 3G out there. We have one of those at work from a Chinese company.